console-subscriber/src/aggregator.rs

```diff
@@ -416,77 +635,129 @@ impl Aggregator {
-        tracing::trace!(
-            ?self.retention,
-            self.has_watchers = has_watchers,
-            "dropping closed tasks..."
-        );
+                self.resource_stats.insert(
+                    id,
+                    ResourceStats {
+                        created_at: Some(at),
+                        ..Default::default()
+                    },
+                );
+            }
 
-        stats.data.retain(|id, (stats, dirty)| {
-            if let Some(closed) = stats.closed_at {
-                let closed_for = now.duration_since(closed).unwrap_or_default();
-                let should_drop =
-                    // if there are any clients watching, retain all dirty tasks regardless of age
-                    (*dirty && has_watchers)
-                    || closed_for > retention;
-                tracing::trace!(
-                    stats.id = ?id,
-                    stats.closed_at = ?closed,
-                    stats.closed_for = ?closed_for,
-                    stats.dirty = *dirty,
-                    should_drop,
+            Event::ResourceOp {
+                metadata,
+                at,
+                resource_id,
+                op_name,
+                op_type,
+                ..
+            } => {
+                match op_type.clone() {
+                    OpType::StateUpdate(attr_updates) => {
+                        if let Some(mut stats) = self.resource_stats.update(&resource_id) {
+                            for upd in attr_updates {
+                                match stats.attributes.get_mut(&upd.name) {
+                                    Some(attr) => attr.update(&upd.val),
+                                    None => {
+                                        stats.attributes.insert(upd.name.clone(), upd.val.into());
+                                    }
+                                }
+                            }
+                        }
+                    }
+                    OpType::Poll {
+                        async_op_id,
+                        task_id,
+                        readiness,
+                    } => {
+                        // TODO: make it less hacky. Ideally we want to store
+                        // these in a vec that gets cleaned periodically and allows
+                        // us to send only the new ones much like the IdData map.
+                        let mut async_op_stats = self.async_op_stats.update_or_default(async_op_id);
+                        async_op_stats.poll_stats.polls += 1;
+                        async_op_stats.latest_poll_op = Some(metadata.into());
+                        if async_op_stats.task_id.is_none() {
+                            async_op_stats.task_id = Some(task_id);
+                        }
+                        if async_op_stats.resource_id.is_none() {
+                            async_op_stats.resource_id = Some(resource_id.clone());
+                        }
```

Comment:
nit, take it or leave it: `Option` has a [`get_or_insert`](https://doc.rust-lang.org/stable/std/option/enum.Option.html#method.get_or_insert) method that we could use here:
```suggestion
                        async_op_stats.task_id.get_or_insert(Some(task_id));
                        async_op_stats.resource_id.get_or_insert(Some(resource_id.clone()));
```
but this might be kind of pointless code golf :woman_shrugging: